### PR TITLE
Separate lexer from parser

### PR DESCRIPTION
**Pull request by thewilsonator**
Use composition instead of inheritance.

Alternative to #9475

cc @jacob-carlborg @andralex

### CODE REVIEW

**Review by jacob-carlborg — APPROVED** [review]

**Review by andralex — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
Test failure:

```
test.d(5):        There's no `static else`, use `else` instead.
Failures:

1) errors supplemental: there's no `static else`, use `else` instead
core.exception.AssertError@/Users/braddr/sandbox/at-client/pull-3648631-Darwin_64_64/dmd/test/unit/parser/diagnostic_reporter.d(67): unittest failure
```

Since it actually prints the error message it looks like the diagnostic reporting is not working properly. It should not print the error message. Instead it should collect it.

File: src/dmd/parse.d [hunk 1]
```diff
@@ -9010,6 +9012,31 @@ final class Parser(AST) : Lexer
...
+        lexer.diagnosticReporter.error(loc, format.ptr, args);
+        va_end(args);
+    }
+    private final void deprecation(const(char)* format, ...)
+    {
+        va_list args;
+        va_start(args, format);
+        lexer.diagnosticReporter.deprecation(lexer.token.loc, format, args);
+        va_end(args);
+    }
```
Comment (jacob-carlborg) [inline src/dmd/parse.d]: Are these three methods necessary? I would think the methods in the lexer would be called automatically due to the `alias this`.

**Comment by andralex** [comment]
I should add that this change is "right" in the sense that `Lexer` should really be a `struct` as its methods are not meant to be overridden. But in terms of separating Lexer from Parser this is a lateral step because it merely replaces one form of subtyping (inheritance) with another (alias this).

File: src/dmd/parse.d [hunk 2]
```diff
@@ -284,6 +284,8 @@ final class Parser(AST) : Lexer
         int inBrackets; // inside [] of array index or slice
         Loc lookingForElse; // location of lonely if looking for an else
     }
+    Lexer lexer;
+    alias lexer this;
```
Comment (WalterBright) [inline src/dmd/parse.d]: I'd rather not use the `alias this`, and simply prefix uses of the lexer with `lexer.`

Comment (thewilsonator) [reply src/dmd/parse.d]: This is replacing the inheritance, the PR will be MUCH bigger without it.

File: src/dmd/parse.d (hunk 1 above)
Comment (thewilsonator) [reply src/dmd/parse.d]: I had hoped that but the seem to pick up `dmd.errors.error` & friends.

Comment (thewilsonator) [reply src/dmd/parse.d]: Note that 
``` 
test.d(5):        There's no `static else`, use `else` instead.
```
 is a supplemental error, which is not defined above.

File: src/dmd/parse.d (hunk 2 above)
Comment (jacob-carlborg) [reply src/dmd/parse.d]: That’s basically what my PR did: https://github.com/dlang/dmd/pull/9475/files

**Comment by jacob-carlborg** [comment]
> I should add that this change is "right" in the sense that Lexer should really be a struct as its methods are not meant to be overridden. But in terms of separating Lexer from Parser this is a lateral step because it merely replaces one form of subtyping (inheritance) with another (alias this).

I see three options:

A. Live with workarounds like `alias this` to be able to make incremental improvements. Eventually they can be removed
B. Accept big PRs that refactors everything at once
C. Do nothing. Which means no progress

**Comment by andralex** [comment]
I think we've been having an "infinite foreplay" pattern. @WalterBright has been overwhelmed stemming a wave of pull requests that he does not agree with on principle, or at best do nothing material but take coding time and review time. Invariably these don't achieve anything but come with the promise that great things will happen once this or that or the other "major but uninteresting" reshuffle will get in. As I mentioned before, in our experience that never happened - any "major but uninteresting" reshuffle reflects the preference of the contributor making it more than good software engineering principles, and does little more than set up a precedent for the next "major but uninteresting" reshuffle, in an endless cycle of "we'll get to work as soon as this is settled".

We need to stop this. @WalterBright (and I for what it's worth) just doesn't buy, and never will, that great people can't get great work done unless we do this one more reshuffle. And that's that. He doesn't buy it and never will. Spending time teaching him the wrongness of his ways is a pure waste of time. Not to mention it's comical that so many try to teach him things and impose their vision over his, but so few wants to learn and do great things aligned with his vision. I humbly suggest those butting heads with @WalterBright and affirming their differences from his ways of doing things to consider where his ways of doing things took his career. Learning from a great hacker is an opportunity that shouldn't be missed.

**Comment by RazvanN7** [comment]
What is the status of this? Do we have an executive decision regarding composition vs. inheritance?

**Pull request closed without merging**